### PR TITLE
Gql

### PR DESCRIPTION
**Pull request by ajeet-smpx**
- added support for `enum` & `variables`.  
  Ex.
  ```js
  const type = VARS.activity.type.COMMENT_ADDED;

  // Using enum
  const tag1 = $gql.tag`
		activity(type: ${$gql.enum(type)}) {
			nodes { id }
		}
  `;
  await $gql.get(tag1, ctx);

  // Using variable
  const tag2 = $gql.tag`
		query($type: ActivityTypeEnum) {
			activity(type: $type) {
				nodes { id }
			}
		}
  `;
  await $gql.get(tag2, ctx, {variables: {type}});
  ```
- https://github.com/smartprix/smartprix/issues/807

### CODE REVIEW

File: src/Gql.js [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+		if (cacheKey && this.cache) {
+			const cached = await this.cache.get(cacheKey);
+			if (cached) return cached;
+		}
+
+		if (!/^\s*query|mutation|subscription/.test(query) && /^\s*[a-zA-Z0-9]/.test(query)) {
+			query = `query { ${query} }`;
+		}
+
+		const result = await graphql(this.defaultSchema, query, null, context, variables);
```
Comment (rohit-gohri) [inline src/Gql.js]: this should be `this.schema`, the selected one via makeSchemaFromConfig options

File: src/Gql.js [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+	let document;
+	try {
+		document = parse(query);
+	}
+	catch (syntaxError) {
+		return Promise.resolve({errors: [syntaxError]});
+	}
+
+	// validate only in developement
+	if (process.env.NODE_ENV !== 'production') {
```
Comment (rohit-gohri) [inline src/Gql.js]: this could also be set through options?

File: src/Gql.js [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+const ONE_DAY = 24 * 3600 * 1000;
+const ENUM_PREFIX = '__ENUM__::';
+// const VAR_PREFIX = '__VAR__::';
+
+// const NO_QUOTES_REGEX = new RegExp(`^(${ENUM_PREFIX}|${VAR_PREFIX})([A-Za-z_]+)$`);
+const ENUM_REGEX = new RegExp(`^${ENUM_PREFIX}([A-Za-z_]+)$`);
+
+function graphql(schemaObj, query, rootValue, contextValue, variableValues) {
+	// we are not using the inbuilt graphql function because it validates
```
Comment (rohit-gohri) [inline src/Gql.js]: move comment to function jsDoc

File: src/Gql.js [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+		if (!result.errors) {
+			if (cacheKey && this.cache) await this.cache.set(cacheKey, result.data, {ttl});
+			return result.data;
+		}
+
+
+		let fields = {};
+		const errors = result.errors;
+
+		// log & format all errors
```
Comment (rohit-gohri) [inline src/Gql.js]: remove comment, as it is no longer logging

File: src/Gql.js [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+		if (!Object.keys(fields).length) {
+			fields = {
+				global: {
+					message: 'Unknown Error',
+					keyword: 'unknown',
+				},
+			};
+		}
+
+		const err = new Error('Error in graphQL api');
```
Comment (rohit-gohri) [inline src/Gql.js]: could mention schema in error message, should be available in `opts.schema` or through `getConfig()`

**Review by rohit-gohri — CHANGES_REQUESTED** [review]
Looks good to me, but this will not be useful for API clients as they won't have a schema to execute.
- We should make the `makeSchemaFromConfig` step optional
- Add an `api` options, something like: `{url: '', auth: '', apiKey: ''}`.
So exec will check if `schema` is available(first priority) or `api`, and if api is available it will send a post request to the api endpoint with the query. This would require adding more options to `exec` to be forwarded to the request, like headers and cookies.

File: src/Gql.js (hunk 1 above)
Comment (ajeet-smpx) [reply src/Gql.js]: `schema.default` is being passed here. So I used the `defaultSchema`.
https://github.com/smartprix/smartprix/blob/0822159767c6bb312d80c563f9079ff20123b19d/src/graphql.js#L121

File: src/Gql.js (hunk 2 above)
Comment (ajeet-smpx) [reply src/Gql.js]: you mean an option of `validateGraphql: boolean` in `Gql.exec`?

Comment (rohit-gohri) [reply src/Gql.js]: yea, or through constructor as this would not be valid for `api` exec

File: index.d.ts [hunk 6]
```diff
@@ -256,4 +256,35 @@ declare module 'gqutils' {
 		parseGraphqlSchemas(): schemaMap;
 		parseGraphqlSchema(schema: string): GraphQLSchema;
 	}
+
+	interface schemaConfigInput extends commonOptions {
+		validateGrqphql?: boolean;
+		cache?: any;
```
Comment (rohit-gohri) [inline index.d.ts]: import Cache from `sm-utils` and use as type, same everywhere else

File: index.d.ts [hunk 7]
```diff
@@ -256,4 +256,35 @@ declare module 'gqutils' {
...
+		formatError?: (error: any) => any;
```
Comment (rohit-gohri) [inline index.d.ts]: Add comment that by default it uses gqutils formatError.
And error type would be `Error`

File: src/Gql.js [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+class ApiError extends Error {}
+class GraphqlError extends Error {}
+
+class Gql {
+	constructor(opts = {}) {
+		if (opts.api) {
+			this.api = _.defaults(opts.api, {
+				endpoint: null,
```
Comment (rohit-gohri) [inline src/Gql.js]: no need to default endpoint. Should actually throw if endpoint is not provided

File: src/Gql.js [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+			.post();
+
+		const result = Str.tryParseJson(response.body);
+
+		if (response.statusCode !== 200) {
+			throw new ApiError(`${response.statusCode}, ${(result && result.errors) || 'Unknown error'}`);
+		}
+
+		if (!result) {
+			throw new ApiError('Invalid result from api');
```
Comment (rohit-gohri) [inline src/Gql.js]: Should add body to error object if it could not be parsed. Or log it

File: src/Gql.js [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+			};
+		}
+
+		const err = new GraphqlError(`[schema:${schemaName || 'default'}] Error in graphQL api`);
+		err.errors = errors;
+		err.fields = fields;
+		throw err;
+	}
+
+	async exec(query, context, {
```
Comment (rohit-gohri) [inline src/Gql.js]: context could also be made optional since api doesn't require it, but don't provide defaults for it

File: src/Gql.js [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+	if (value == null) return null;
+
+	if (typeof value === 'number') return String(value);
+	if (typeof value === 'string') return convertStrToGqlArg(value);
+	if (_.isPlainObject(value)) return `{${convertObjToGqlArg(value)}}`;
+
+	return JSON.stringify(value);
+}
+
+class ApiError extends Error {}
```
Comment (rohit-gohri) [inline src/Gql.js]: These could also be exported. Would be helpful while logging to check if an error is instance of any of these

**Review by rohit-gohri — CHANGES_REQUESTED** [review]
Will we be able to log all of the cases from outside?

**Review by rohit-gohri — APPROVED** [review]

**Comment by hit12369** [comment]
Rather than using `ENUM_PREFIX`, enum should be an instance of a different class.

```js
class GqlEnum {
  constructor(val) { this.val = val; }
  toString() { return this.val; }
}

function enum(val) { return new GqlEnum(val); }

if (val instanceof GqlEnum) return val.toString();
```

**Pull request merged**